src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/level/MixinLevel.java

```diff
@@ -1,44 +1,209 @@
+    @Shadow public abstract long getDayTime();
 
-    // getBlockState - mixin
-
-    // getBlockEntity - mixin
-
-    // setBlockEntity - mixin
-
-    // removeBlockEntity - mixin
-
-    // isLoaded - mixin
+    protected boolean cc_isCubic;
 
-    // loadedAndEntityCanStandOnFace - mixin
+    @Override
+    public void cc_setCubic() {
+        cc_isCubic = true;
+    }
 
-    // getChunkForCollisions - mixin
+    @Override public boolean cc_isCubic() {
+        return cc_isCubic;
+    }
 
-    // blockEntityChanged - mixin
+    public LevelCube cc_getCubeAt(BlockPos pos) {
+        return (LevelCube)((CubicLevelReader)this).cc_getCube(Coords.blockToCube(pos.getX()), Coords.blockToCube(pos.getY()),
+            Coords.blockToCube(pos.getZ()));
+    }
 
-    // getCurrentDifficultyAt - mixin
+    public LevelCube cc_getCube(int x, int y, int z) {
+        return (LevelCube)this.cc_getCube(x, y, z, ChunkStatus.FULL);
+    }
 
-    // TODO: Phase 3 low priority: Add a method to modify isOutsideSpawnableHeight to go from -30 mil to 30 mil, not -20 mil to 20 mil
+    @Nullable
+    @Override
+    public CubeAccess cc_getCube(int cubeX, int cubeY, int cubeZ, ChunkStatus requiredStatus, boolean nonnull) {
+        CubeAccess cubeaccess = ((CubicChunkSource)this.getChunkSource()).cc_getCube(cubeX, cubeY, cubeZ, requiredStatus, nonnull);
+        if (cubeaccess == null && nonnull) {
+            throw new IllegalStateException("Should always be able to create a cube!");
+        } else {
+            return cubeaccess;
+        }
+    }
+
+    // setBlock
+    @WrapOperation(method = "setBlock(Lnet/minecraft/core/BlockPos;Lnet/minecraft/world/level/block/state/BlockState;II)Z", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level"
+        + "/Level;getChunkAt(Lnet/minecraft/core/BlockPos;)Lnet/minecraft/world/level/chunk/LevelChunk;"))
+    private LevelChunk cc_replaceLevelChunkInGetChunkAt(Level level, BlockPos pos, Operation<LevelChunk> original, @Share("levelCube") LocalRef<LevelCube> levelCubeLocalRef) {
+        if(cc_isCubic) {
+            levelCubeLocalRef.set(this.cc_getCubeAt(pos));
+            return null;
+        }
+        return original.call(level, pos);
+    }
+
+    @WrapOperation(method = "setBlock(Lnet/minecraft/core/BlockPos;Lnet/minecraft/world/level/block/state/BlockState;II)Z", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/chunk/LevelChunk;setBlockState(Lnet/minecraft/core/BlockPos;Lnet/minecraft/world/level/block/state/BlockState;Z)Lnet/minecraft/world/level/block/state/BlockState;"))
+    private BlockState cc_replaceLevelChunkInSetBlockState(LevelChunk levelChunk, BlockPos pos, BlockState state, boolean isMoving, Operation<BlockState> original,
+                                                           @Share("levelCube") LocalRef<LevelCube> levelCubeLocalRef) {
+        if(cc_isCubic) {
+            levelCubeLocalRef.get().setBlockState(pos, state, isMoving);
+            return null;
```

Comment:
This should return the value from the `setBlockState` call above.
`Level#setBlock` returns the block that was replaced